### PR TITLE
Replica Count Validation when awareness replica balance is enabled

### PR DESCRIPTION
**Pull request by gbbafna**
Signed-off-by: Gaurav Bafna <gbbafna@amazon.com>

*Issue #, if available:*

https://github.com/opensearch-project/index-management/issues/381

*Description of changes:*

*CheckList:*
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/index-management/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by gbbafna** [comment]
@bowenlan-amzn  : This change depends on ongoing changes in OpenSearch . Is there a mechanism to run integ tests with local opensearch zip ? 

I did try copying over the zip to `~/.gradle/caches/modules-2/files-2.1/org.opensearch.distribution.integ-test-zip/opensearch/2.1.0-SNAPSHOT/66102a4554a8e60c0c9ae5c82456ca04d850069e/opensearch-2.1.0-SNAPSHOT.zip` , but still the integ test is failing over.

**Comment by bowenlan-amzn** [comment]
> @bowenlan-amzn : This change depends on ongoing changes in OpenSearch . Is there a mechanism to run integ tests with local opensearch zip ?
> 
> I did try copying over the zip to `~/.gradle/caches/modules-2/files-2.1/org.opensearch.distribution.integ-test-zip/opensearch/2.1.0-SNAPSHOT/66102a4554a8e60c0c9ae5c82456ca04d850069e/opensearch-2.1.0-SNAPSHOT.zip` , but still the integ test is failing over.

Didn't try to do such thing before. I think @saratvemulapalli may know the way to do this.

**Comment by saratvemulapalli** [comment]
@gbbafna sure. Yeah you should be able to run via custom distribution. 
Something like: 
```
./gradlew build -PcustomDistributionUrl=https://ci.opensearch.org/ci/dbc/distribution-build-opensearch/2.1.1/latest/darwin/x64/tar/builds/opensearch/dist/opensearch-min-2.1.1-darwin-x64.tar.gz
```
Take a look at: https://github.com/opensearch-project/opensearch-plugins/blob/main/TESTING.md#distribution-download-plugin

**Comment by gbbafna** [comment]
OpenSearch changes are merged. We can trigger rerun of the jobs once we have a snapshot build available .

**Comment by gbbafna** [comment]
@bowenlan-amzn, @saratvemulapalli   : The opensearch changes are merged into main as well as `2.x` branch. How do we get this changes to build with that snapshot ?

File: src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/IndexPolicyActionIT.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            id = policyID,
+            description = "$testIndexName description",
+            schemaVersion = 1L,
+            lastUpdatedTime = Instant.now().truncatedTo(ChronoUnit.MILLIS),
+            errorNotification = randomErrorNotification(),
+            defaultState = states[0].name,
+            states = states
+        )
+        Assert.assertThrows(
+            ResponseException::class.java
```
Comment (bowenlan-amzn) [inline src/test/kotlin/org/opensearch/indexmanagement/indexstatemanagement/action/IndexPolicyActionIT.kt:74]: nit: better if we can check the validation error message

**Comment by saratvemulapalli** [comment]
@gbbafna Hmm what are you really looking for ?

**Comment by gbbafna** [comment]
> @gbbafna Hmm what are you really looking for ?
> If the changes are in 2.x, I would build this plugin with 2.2 and consume the latest snapshot of OpenSearch 2.2.

2.2 branch in OpenSearch is not cut yet . It will be cut on 04/04 and there will be code freeze after that . So I am bit confused about the way forward here. 

I have built this plugin locally and tested it . It is only on github that all the checks are failing.

**Comment by saratvemulapalli** [comment]
@gbbafna you are correct. As soon as the version is released there shouldn't be any development on that version. 
The version has to be bumped. So yes, ISM should move to 2.1.1.

**Comment by gbbafna** [comment]
I tried to bump the version to 2.1.1, but it failed due to absence of notification plugin , similar to as observed in https://github.com/opensearch-project/index-management/pull/443 .

I will wait for 2.2.0 PR https://github.com/opensearch-project/index-management/pull/443 to go through and then re run gradle checks.

**Review by bowenlan-amzn — DISMISSED** [review]

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt [hunk 2]
```diff
@@ -103,6 +109,23 @@ class TransportIndexPolicyAction @Inject constructor(
...
+                        if (error.isPresent) {
+                            val ex = ValidationException()
+                            ex.addValidationError(error.get())
+                            actionListener.onFailure(ex)
+                        }
+                    }
+                }
+            }
+        }
+
```
Comment (Angie-Zhang) [inline src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:128]: To keep coding style consistency, we can write this method as:

```
private fun validate() {
    request.policy.states.forEach { state ->
        state.actions.forEach { action ->
            if (action is ReplicaCountAction ) {
                val error = awarenessReplicaBalance.validate(action.numOfReplicas)
                if (error.isPresent) {
                    val ex = ValidationException()
                    ex.addValidationError(error.get())
                    actionListener.onFailure(ex)
                }
            }
        }
    }
}
```

**Review by Angie-Zhang — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt (hunk 2 above)
Comment (gbbafna) [reply src/main/kotlin/org/opensearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/TransportIndexPolicyAction.kt:127]: Thanks, done.

**Review by bowenlan-amzn — APPROVED** [review]

**Review by Angie-Zhang — APPROVED** [review]

**Pull request merged**